File: pml/lexer.py
"""PML lexer — hand-written scanner for S-expression tokens."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto

from pml.errors import PMLSyntaxError


class TokenType(Enum):
    INTEGER = auto()
    FLOAT = auto()
    STRING = auto()
    SYMBOL = auto()
    BOOLEAN = auto()
    KEYWORD = auto()
    LPAREN = auto()
    RPAREN = auto()
    QUOTE = auto()
    QUASIQUOTE = auto()
    UNQUOTE = auto()
    UNQUOTE_SPLICE = auto()
    EOF = auto()


@dataclass(slots=True)
class Token:
    type: TokenType
    value: str
    line: int
    column: int

    def __repr__(self) -> str:
        return f"Token({self.type.name}, {self.value!r}, {self.line}:{self.column})"


# Characters that terminate a symbol (besides whitespace)
_DELIMITERS = set("()\"`; \t\n\r")

# Characters that cannot appear in symbols
_SYMBOL_SPECIAL = set("-?*+<=>!_/")


class Lexer:
    """Tokenize PML source code into a list of Tokens."""

    def __init__(self, source: str, filename: str = "<stdin>") -> None:
        self.source = source
        self.filename = filename
        self.pos = 0
        self.line = 1
        self.column = 1

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def tokenize(self) -> list[Token]:
        """Scan the entire source and return a list of tokens."""
        tokens: list[Token] = []
        while self.pos < len(self.source):
            self._skip_whitespace_and_comments()
            if self.pos >= len(self.source):
                break
            token = self._read_token()
            if token is not None:
                tokens.append(token)
        tokens.append(Token(TokenType.EOF, "", self.line, self.column))
        return tokens

    # ------------------------------------------------------------------
    # Whitespace & comments
    # ------------------------------------------------------------------

    def _skip_whitespace_and_comments(self) -> None:
        while self.pos < len(self.source):
            ch = self.source[self.pos]
            if ch in " \t\n\r":
                self._advance()
            elif ch == ";":
                # Line comment — skip to end of line
                while self.pos < len(self.source) and self.source[self.pos] != "\n":
                    self._advance()
            else:
                break

    # ------------------------------------------------------------------
    # Token reading
    # ------------------------------------------------------------------

    def _read_token(self) -> Token | None:
        ch = self.source[self.pos]
        line, col = self.line, self.column

        # Parentheses
        if ch == "(":
            self._advance()
            return Token(TokenType.LPAREN, "(", line, col)
        if ch == ")":
            self._advance()
            return Token(TokenType.RPAREN, ")", line, col)

        # String literal
        if ch == '"':
            return self._read_string()

        # Syntax sugar prefixes
        if ch == "'":
            self._advance()
            return Token(TokenType.QUOTE, "'", line, col)
        if ch == "`":
            self._advance()
            return Token(TokenType.QUASIQUOTE, "`", line, col)
        if ch == ",":
            self._advance()
            # Check for ,@  (unquote-splicing)
            if self.pos < len(self.source) and self.source[self.pos] == "@":
                self._advance()
                return Token(TokenType.UNQUOTE_SPLICE, ",@", line, col)
            return Token(TokenType.UNQUOTE, ",", line, col)

        # Number or symbol
        return self._read_atom()

    # ------------------------------------------------------------------
    # Atoms
    # ------------------------------------------------------------------

    def _read_string(self) -> Token:
        """Read a double-quoted string with escape sequences."""
        line, col = self.line, self.column
        self._advance()  # skip opening "
        chars: list[str] = []
        while self.pos < len(self.source):
            ch = self.source[self.pos]
            if ch == "\\":
                self._advance()
                if self.pos >= len(self.source):
                    raise PMLSyntaxError(
                        "Unterminated string escape",
                        line=line,
                        column=col,
                        filename=self.filename,
                    )
                esc = self.source[self.pos]
                escape_map = {"n": "\n", "t": "\t", "\\": "\\", '"': '"'}
                chars.append(escape_map.get(esc, esc))
                self._advance()
            elif ch == '"':
                self._advance()  # skip closing "
                return Token(TokenType.STRING, "".join(chars), line, col)
            else:
                chars.append(ch)
                self._advance()
        raise PMLSyntaxError(
            "Unterminated string literal",
            line=line,
            column=col,
            filename=self.filename,
        )

    def _read_atom(self) -> Token:
        """Read a number, boolean, keyword, or symbol."""
        line, col = self.line, self.column
        start = self.pos
        while self.pos < len(self.source) and self.source[self.pos] not in _DELIMITERS:
            self._advance()
        text = self.source[start : self.pos]

        if not text:
            raise PMLSyntaxError(
                f"Unexpected character: {self.source[self.pos]!r}",
                line=line,
                column=col,
                filename=self.filename,
            )

        # Booleans
        if text == "#t":
            return Token(TokenType.BOOLEAN, "#t", line, col)
        if text == "#f":
            return Token(TokenType.BOOLEAN, "#f", line, col)

        # Keyword (starts with ':')
        if text.startswith(":"):
            return Token(TokenType.KEYWORD, text[1:], line, col)

        # Number: try integer then float
        if self._looks_like_number(text):
            if "." in text:
                return Token(TokenType.FLOAT, text, line, col)
            return Token(TokenType.INTEGER, text, line, col)

        # Symbol
        return Token(TokenType.SYMBOL, text, line, col)

    @staticmethod
    def _looks_like_number(text: str) -> bool:
        """Check whether text represents a number."""
        if not text:
            return False
        # Strip leading sign
        t = text
        if t[0] in "+-":
            t = t[1:]
        if not t:
            return False
        # Must start with digit or dot-digit
        if t[0] == ".":
            t = t[1:]
            if not t:
                return False
        # All remaining chars must be digits (with at most one dot)
        dot_seen = False
        for ch in t:
            if ch == ".":
                if dot_seen:
                    return False
                dot_seen = True
            elif not ch.isdigit():
                return False
        return True

    # ------------------------------------------------------------------
    # Cursor helpers
    # ------------------------------------------------------------------

    def _advance(self) -> None:
        """Move one character forward, tracking line/column."""
        if self.pos < len(self.source):
            if self.source[self.pos] == "\n":
                self.line += 1
                self.column = 1
            else:
                self.column += 1
            self.pos += 1

File: pml-cpp/src/pml/frontend/lexer.cpp
// ═══════════════════════════════════════════════════════════════════════════════
// PML Lexer — Hand-written Scanner for S-expression Tokens
// ═══════════════════════════════════════════════════════════════════════════════
//
// Matches the Python PML lexer (pml/lexer.py) semantics exactly.
// ═══════════════════════════════════════════════════════════════════════════════

#include "lexer.h"

#include <cctype>
#include <format>
#include <utility>

namespace pml {

// ═══════════════════════════════════════════════════════════════════════════════
// Constructor
// ═══════════════════════════════════════════════════════════════════════════════

Lexer::Lexer(std::string source_, std::string filename_)
    : source(std::move(source_))
    , filename(std::move(filename_)) {}

// ═══════════════════════════════════════════════════════════════════════════════
// Public API
// ═══════════════════════════════════════════════════════════════════════════════

auto Lexer::tokenize() -> Result<std::vector<Token>> {
    std::vector<Token> tokens;
    while (pos < source.size()) {
        skip_whitespace_and_comments();
        if (pos >= source.size()) {
            break;
        }
        auto tok = read_token();
        if (!tok) {
            return std::unexpected(tok.error());
        }
        tokens.push_back(std::move(*tok));
    }
    tokens.push_back(Token{TokenType::END_OF_FILE, "", line, column});
    return tokens;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Whitespace & comments
// ═══════════════════════════════════════════════════════════════════════════════

void Lexer::skip_whitespace_and_comments() {
    while (pos < source.size()) {
        const char ch = source[pos];
        if (ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r') {
            advance();
        } else if (ch == ';') {
            // Line comment — skip to end of line
            while (pos < source.size() && source[pos] != '\n') {
                advance();
            }
        } else {
            break;
        }
    }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Token reading dispatcher
// ═══════════════════════════════════════════════════════════════════════════════

auto Lexer::read_token() -> Result<Token> {
    const char ch = source[pos];
    const int cur_line = line;
    const int cur_col = column;

    // Parentheses
    if (ch == '(') {
        advance();
        return Token{TokenType::LPAREN, "(", cur_line, cur_col};
    }
    if (ch == ')') {
        advance();
        return Token{TokenType::RPAREN, ")", cur_line, cur_col};
    }

    // String literal
    if (ch == '"') {
        return read_string();
    }

    // Syntax sugar prefixes
    if (ch == '\'') {
        advance();
        return Token{TokenType::QUOTE, "'", cur_line, cur_col};
    }
    if (ch == '`') {
        advance();
        return Token{TokenType::QUASIQUOTE, "`", cur_line, cur_col};
    }
    if (ch == ',') {
        advance();
        // Check for ,@  (unquote-splicing)
        if (pos < source.size() && source[pos] == '@') {
            advance();
            return Token{TokenType::UNQUOTE_SPLICE, ",@", cur_line, cur_col};
        }
        return Token{TokenType::UNQUOTE, ",", cur_line, cur_col};
    }

    // Number or symbol
    return read_atom();
}

// ═══════════════════════════════════════════════════════════════════════════════
// String literal reader
// ═══════════════════════════════════════════════════════════════════════════════

auto Lexer::read_string() -> Result<Token> {
    const int str_line = line;
    const int str_col = column;
    advance();  // skip opening "

    std::string chars;
    while (pos < source.size()) {
        const char ch = source[pos];
        if (ch == '\\') {
            advance();
            if (pos >= source.size()) {
                return std::unexpected(syntax_error(
                    SourceLocation{str_line, str_col, filename},
                    "Unterminated string escape"
                ));
            }
            const char esc = source[pos];
            switch (esc) {
                case 'n':  chars.push_back('\n'); break;
                case 't':  chars.push_back('\t'); break;
                case '\\': chars.push_back('\\'); break;
                case '"':  chars.push_back('"');  break;
                default:   chars.push_back(esc);  break;  // unknown escape: keep as-is
            }
            advance();
        } else if (ch == '"') {
            advance();  // skip closing "
            return Token{TokenType::STRING, std::move(chars), str_line, str_col};
        } else {
            chars.push_back(ch);
            advance();
        }
    }

    return std::unexpected(syntax_error(
        SourceLocation{str_line, str_col, filename},
        "Unterminated string literal"
    ));
}

// ═══════════════════════════════════════════════════════════════════════════════
// Atom reader (number, boolean, keyword, or symbol)
// ═══════════════════════════════════════════════════════════════════════════════

auto Lexer::read_atom() -> Result<Token> {
    const int atom_line = line;
    const int atom_col = column;
    const size_t start = pos;

    // Characters that terminate an atom (besides whitespace)
    // Must match the Python _DELIMITERS set: set("()\"`; \t\n\r")
    const auto is_delimiter = [](char ch) -> bool {
        return ch == '(' || ch == ')' || ch == '"' || ch == '`' || ch == ';'
            || ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r';
    };

    while (pos < source.size() && !is_delimiter(source[pos])) {
        advance();
    }

    const std::string text = source.substr(start, pos - start);

    if (text.empty()) {
        return std::unexpected(syntax_error(
            SourceLocation{atom_line, atom_col, filename},
            std::format("Unexpected character: {}", source[pos])
        ));
    }

    // Booleans
    if (text == "#t") {
        return Token{TokenType::BOOLEAN, "#t", atom_line, atom_col};
    }
    if (text == "#f") {
        return Token{TokenType::BOOLEAN, "#f", atom_line, atom_col};
    }

    // Keyword (starts with ':')
    if (text[0] == ':') {
        return Token{TokenType::KEYWORD, text.substr(1), atom_line, atom_col};
    }

    // Number: try integer then float
    if (looks_like_number(text)) {
        if (text.find('.') != std::string::npos) {
            return Token{TokenType::FLOAT, text, atom_line, atom_col};
        }
        return Token{TokenType::INTEGER, text, atom_line, atom_col};
    }

    // Symbol
    return Token{TokenType::SYMBOL, text, atom_line, atom_col};
}

// ═══════════════════════════════════════════════════════════════════════════════
// Number detection
// ═══════════════════════════════════════════════════════════════════════════════

auto Lexer::looks_like_number(const std::string& text) -> bool {
    if (text.empty()) {
        return false;
    }

    // Strip leading sign
    size_t i = 0;
    if (text[0] == '+' || text[0] == '-') {
        i = 1;
    }
    if (i >= text.size()) {
        return false;
    }

    // Must start with digit or dot-digit
    if (text[i] == '.') {
        ++i;
        if (i >= text.size()) {
            return false;
        }
    }

    // All remaining chars must be digits (with at most one dot)
    bool dot_seen = false;
    for (/* i already set */; i < text.size(); ++i) {
        const char ch = text[i];
        if (ch == '.') {
            if (dot_seen) {
                return false;
            }
            dot_seen = true;
        } else if (!std::isdigit(static_cast<unsigned char>(ch))) {
            return false;
        }
    }
    return true;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Cursor helpers
// ═══════════════════════════════════════════════════════════════════════════════

void Lexer::advance() {
    if (pos < source.size()) {
        if (source[pos] == '\n') {
            ++line;
            column = 1;
        } else {
            ++column;
        }
        ++pos;
    }
}

}  // namespace pml
